File: reverse-linked-list/reverse-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if(head==nullptr) return nullptr;
        if(head->next == nullptr) return head;
        ListNode* prev=nullptr;
        ListNode* curr=head;
        // ListNode* next=curr->next;
        return reverse(prev,curr);
        
        
    }
    ListNode* reverse(ListNode* prev,ListNode* curr){
        if(curr==nullptr) return prev;
        if(curr->next==nullptr) {
            curr->next = prev;
            return curr;
        }
        ListNode* next=curr->next;
        ListNode* second=next->next;
        curr->next->next=curr;
        curr->next=prev;
        prev=next;
        curr=second;
        return reverse(prev,curr);
        
    }
    
};

File: reverse-linked-list/reverse-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        if(head==null || head.next==null) return head;
        ListNode dummy=null;
        //  dummy.next=head;
        ListNode prev=dummy;
        ListNode curr=head;
        ListNode next=head.next;
        while(curr!=null){
            next=curr.next;
            curr.next=prev;
            prev=curr;
            curr=next;
        }
        return prev;
        
        
    }
}